lib/Mojolicious/Plugin/OpenAPI/Cors.pm

```diff
@@ -48,63 +169,187 @@ Mojolicious::Plugin::OpenAPI::Cors - OpenAPI plugin for Cross-Origin Resource Sh
 
 =head1 SYNOPSIS
 
-  package MyApplication::Controller::User;
+=head2 Application
+
+Set L</add_preflighted_routes> to 1, if you want "Preflighted" CORS requests to
+be sent to your already existing actions.
+
+  $app->plugin("OpenAPI" => {add_preflighted_routes => 1});
+
+=head2 Simple exchange
+
+The following example will automatically set default CORS response headers
+after validating the request against L</openapi_cors_allow_origins>:
+
+  package MyApp::Controller::User;
 
   sub get_user {
+    my $c = shift->openapi->cors_exchange->openapi->valid_input or return;
+
+    # Will only run this part if both the cors_exchange and valid_input was
+    # successful.
+    $c->render(openapi => {user => {}});
+  }
 
+=head2 Custom exchange
+
+If you need full control, you must pass a callback to
+L</openapi.cors_exchange>:
+
+  package MyApp::Controller::User;
+
+  sub get_user {
     # Validate incoming CORS request with _validate_cors()
-    my $c = shift->openapi->cors_simple("_validate_cors")->openapi->valid_input or return;
+    my $c = shift->openapi->cors_exchange("_validate_cors")->openapi->valid_input or return;
 
+    # Will only run this part if both the cors_exchange and valid_input was
+    # successful.
     $c->render(openapi => {user => {}});
   }
 
+  # This method must return undef on success. Any true value will be used as
+  # an error.
   sub _validate_cors {
-    my ($c, $args) = @_;
+    my ($c, $params) = @_;
+
+    # $params->{type} is set to "preflighted", "real" or "simple"
+    $c->app->log->debug("Got CORS $params->{type} request");
+
+    # The following "Origin" header check is the same for both simple and
+    # preflighted.
+    return "/Origin" unless $params->{origin} =~ m!^https?://whatever.example.com!;
 
-    # Check the origin of the request
-    if ($args->{origin} =~ m!^https?://whatever.example.com!) {
+    # The following checks are only valid if preflighted...
 
-      # Setting the "Access-Control-Allow-Origin" will mark this request as valid
-      $c->res->headers->header("Access-Control-Allow-Origin" => $args->{origin});
+    # Check the Access-Control-Request-Headers header
+    return "Bad stuff." if $params->{headers} and $params->{headers} =~ /X-No-Can-Do/;
+
+    # Check the Access-Control-Request-Method header
+    return "Not cool." if $params->{method} and $params->{method} eq "delete";
```

Comment:
I assume the "delete" should be "DELETE"? If not, I'd prefer it be capitalized to agree with `$c->method`.